pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		//TODO: see if we can add a test for valid but error throwing service, its difficult right now because synCService() currently runtime.HandleError
+		{
+			testName: "if valid service",
+			key:      "external-balancer",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for _, tst := range testCases {
+		var printError bool
+		tst.updateFn()
+		obtainedErr := controller.syncService(tst.key)
+		if obtainedErr != tst.expectedErr {
+			printError = true
+			if obtainedErr != nil && tst.expectedErr != nil {
+				if obtainedErr.Error() == tst.expectedErr.Error() {
+					//Check if the actual error matches, if yes then test PASSed
+					printError = false
+				}
+			}
+		}
+		if printError {
+			t.Errorf("%v Expected=%v Obtained=%v", tst.testName, tst.expectedErr, obtainedErr)
+		}
+	}
+}
+
+func TestDeleteService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
+	}{
+		{
+			testName: "If an invalid service is deleted",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				expectedError := "Service external-balancer not in cache even though the watcher thought it was. Ignoring the deletion."
+				if srvErr == nil || srvErr.Error() != expectedError {
+					//cannot be nil or Wrong error message
+					return fmt.Errorf("Expected=%v Obtained=%v", expectedError, srvErr)
+				}
+
+				if retryDuration != doNotRetry {
+					//Retry duration should match
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+
+				return nil
+			},
+		},
+		{
+			testName: "If cloudprovided failed to delete the service",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+				cloud.Err = fmt.Errorf("Error Deleting the Loadbalancer")
+
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				expectedError := "Error Deleting the Loadbalancer"
+
+				if srvErr == nil || srvErr.Error() != expectedError {
+					return fmt.Errorf("Expected=%v Obtained=%v", expectedError, srvErr)
+				}
+
+				if retryDuration != minRetryDelay {
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", minRetryDelay, retryDuration)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "If delete was successful",
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
```

Comment:
Seems like newController() is needed for all test cases. Why not move it out of the function and pass it in asan argument?